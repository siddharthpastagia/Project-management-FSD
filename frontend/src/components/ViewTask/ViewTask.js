import React, { useState, useEffect } from "react";
import { withRouter, useHistory } from "react-router-dom";
import {
  Container,
  Row,
  Col,
  FormLabel,
  InputGroup,
  FormControl,
  Button,
  ListGroup,
  ListGroupItem,
  Alert
} from "react-bootstrap";
import moment from "moment";
import * as _ from "lodash";
import SearchModal from "../common/SearchModal";
import {
  getAllTasksByProjectId,
  updateTaskAsComplete,
  getAllProject,
  getTaskById
} from "../../api/Api";

const ViewTask = () => {
  let [tasks, setTasks] = useState([]);
  let [statusMessage, setStatusMessage] = useState({
    show: false,
    message: "",
    variant: ""
  });

  //////////////////////////////////////////
  //Serach Project
  let history = useHistory();
  let [project, setProject] = useState("");
  let [projectList, setProjectList] = useState([]);
  let [showProjectModal, setShowProjectModal] = useState(false);
  let [sortMode, setSortMode] = useState(false);
  let [startDateSort, setStartDateSort] = useState(false);
  let [endDateSort, setEndDateSort] = useState(false);
  let [prioritySort, setPrioritySort] = useState(false);
  let [taskCompleted, setTaskCompleted] = useState(false);
  const searchProject = async () => {
    setProjectList(await getAllProject());
    setShowProjectModal(true);
  };
  function onCloseProjectModal(val) {
    setShowProjectModal(false);
  }
  function onSearchProject(project) {
    setProject(project);
  }

  async function fetchTasksByProjectId() {
    if (project) {
      try {
        setTasks(await getAllTasksByProjectId(project));
      } catch (err) {
        setStatusMessage({
          ...statusMessage,
          show: true,
          message: err,
          variant: "danger"
        });
      }
      setSortMode(false);
    }
  }

  function formatDate(date) {
    return moment(date).format("YYYY-MM-DD");
  }

  // edit project

  const editTask = async task => {
    const currentTask = await getTaskById(task._id);
    history.push("add-task", currentTask);
  };
  // sorting

  const handleSort = field => {
    setSortMode(true);
    const sortByField = _.sortBy(tasks, field);
    setTasks(sortByField);
  };
  useEffect(() => {
    fetchTasksByProjectId();
  }, [project]);
  return (
    <>
      <Container>
        {taskCompleted && (
          <Alert variant="info" className="mt-3">
            Selected task has been Completed.
          </Alert>
        )}
        <Row>
          <Col className="mt-3">
            <InputGroup>
              <FormControl
                required
                readOnly
                name="project"
                value={project.projectName}
              ></FormControl>
              <InputGroup.Append>
                <Button variant="outline-primary" onClick={searchProject}>
                  Search project
                </Button>
              </InputGroup.Append>
            </InputGroup>
          </Col>
          <SearchModal
            id="searchProject"
            heading="Serach Project"
            showModal={showProjectModal}
            onCloseModal={onCloseProjectModal}
            data={projectList}
            columns={[
              { dataField: "_id", hidden: true },
              { dataField: "projectName", text: "Project List" }
            ]}
            onSearch={onSearchProject}
          />
        </Row>
        <Row className="mt-3 mb-3">
          <Col xs={12} sm={6}>
            Sort Task By:
            <Button
              variant="outline-primary"
              className={sortMode && startDateSort ? "active ml-2" : "ml-2"}
              onClick={() => {
                handleSort("startDate");
                setStartDateSort(true);
                setEndDateSort(false);
                setPrioritySort(false);
              }}
            >
              Start Date
            </Button>
            <Button
              variant="outline-primary"
              className={sortMode && endDateSort ? "active ml-2" : "ml-2"}
              onClick={() => {
                handleSort("endDate");
                setStartDateSort(false);
                setEndDateSort(true);
                setPrioritySort(false);
              }}
            >
              End Date
            </Button>
            <Button
              variant="outline-primary"
              className={sortMode && prioritySort ? "active ml-2" : "ml-2"}
              onClick={() => {
                handleSort("priority");
                setStartDateSort(false);
                setEndDateSort(false);
                setPrioritySort(true);
              }}
            >
              Priority
            </Button>
          </Col>
        </Row>
        {tasks.length > 0 ? (
          <Row>
            <Col>
              <table className="table table-bordered mt-5">
                <thead>
                  <tr>
                    <th>Task</th>
                    <th>Parent</th>
                    <th>Priority</th>
                    <th>Start</th>
                    <th>End</th>
                    <th></th>
                  </tr>
                </thead>
                <tbody>
                  {tasks.map(task => {
                    return (
                      <tr>
                        <td>{task.taskName}</td>
                        <td>{task.parentTask && task.parentTask.taskName}</td>
                        <td>{task.priority}</td>
                        <td>{formatDate(task.startDate)}</td>
                        <td>{formatDate(task.endDate)}</td>
                        <td>
                          <Button
                            variant="outline-primary"
                            onClick={() => editTask(task)}
                          >
                            Edit
                          </Button>
                          <Button
                            variant="outline-primary"
                            className="ml-2"
                            disabled={
                              taskCompleted || task.status === "Completed"
                            }
                            onClick={async () => {
                              await updateTaskAsComplete(task);
                              setTaskCompleted(true);
                            }}
                          >
                            End Task
                          </Button>
                        </td>
                      </tr>
                    );
                  })}
                </tbody>
              </table>
            </Col>
          </Row>
        ) : (
          <Alert variant="warning">
            No task available for selected project, please select another
            project from the list
          </Alert>
        )}
      </Container>
    </>
  );
};

export default withRouter(ViewTask);
