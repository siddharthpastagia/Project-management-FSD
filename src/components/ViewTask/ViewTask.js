import React from "react";

export const ViewTask = () => {
  return <>This is View Task screen</>;
};
