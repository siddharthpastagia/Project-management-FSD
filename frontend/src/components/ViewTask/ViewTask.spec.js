import React from "react";
import { MemoryRouter as Router } from "react-router-dom";
import {
  render,
  cleanup,
  wait,
  fireEvent,
  getByTestId
} from "@testing-library/react";
import ViewTask from "./ViewTask";
import axios from "axios";
jest.mock("axios");

describe("View Task Component", () => {
  afterEach(cleanup);
  it("Should render ViewTask component", () => {
    const { container } = render(
      <Router>
        <ViewTask />
      </Router>
    );
    expect(container).toMatchSnapshot();
  });

  it("should not get any task before project selection", () => {
    const { container, getByText } = render(
      <Router>
        <ViewTask />
      </Router>
    );
    expect(
      getByText(
        "No task available for selected project, please select another project from the list"
      )
    );
  });

  // it("should load projects on search icon click", () => {
  //   let projectList = {
  //     data: [
  //       {
  //         _id: "5def0350ba3cb90f6c91ccc3",
  //         projectName: "New",
  //         priority: 23,
  //         dateRequired: false,
  //         task: [
  //           {
  //             _id: "5df2b9501eed66114ccd3f68",
  //             project: "5def0350ba3cb90f6c91ccc3",
  //             taskName: "Amit Task",
  //             isParentTask: true,
  //             __v: 0,
  //             status: "Completed",
  //             id: "5df2b9501eed66114ccd3f68"
  //           }
  //         ],
  //         id: "5def0350ba3cb90f6c91ccc3"
  //       },
  //       {
  //         _id: "5def099bba3cb90f6c91ccc5",
  //         projectName: "dege",
  //         priority: 54,
  //         dateRequired: false,
  //         __v: 0,
  //         task: [
  //           {
  //             _id: "5df2bbb61eed66114ccd3f6b",
  //             project: "5def099bba3cb90f6c91ccc5",
  //             taskName: "jooo",
  //             isParentTask: true,
  //             status: "Completed",
  //             id: "5df2bbb61eed66114ccd3f6b"
  //           }
  //         ],
  //         id: "5def099bba3cb90f6c91ccc5"
  //       }
  //     ]
  //   };
  //   const { container, getByTestId, getByText } = render(
  //     <Router>
  //       <ViewTask />
  //     </Router>
  //   );

  //   const searchModalIcon = getByTestId("searchIcon");
  //   fireEvent.click(searchModalIcon);
  //   // const getAllProjects = jest.fn();
  //   // axios.get.mockImplementationOnce(() =>
  //   //   Promise.resolve({ data: projectList })
  //   // );
  //   // expect(getByText("Search Project"));
  // });
});
