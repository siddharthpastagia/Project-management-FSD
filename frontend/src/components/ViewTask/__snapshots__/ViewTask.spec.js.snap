// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Task Component Should render ViewTask component 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <h3
          class="mt-3"
        >
          View Tasks
        </h3>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="mt-3 col"
      >
        <label
          class="form-label"
        >
          Search Project
        </label>
        <div
          class="input-group"
        >
          <input
            class="form-control"
            name="project"
            readonly=""
            required=""
            value=""
          />
          <div
            class="input-group-append"
          >
            <button
              class="btn btn-outline-primary"
              data-testid="searchIcon"
              type="button"
            >
              <i
                class="fa fa-search"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="mt-3 mb-3 row"
    >
      <div
        class="col-sm-6 col-12"
      >
        Sort Task By:
        <button
          class="ml-2 btn btn-outline-primary"
          type="button"
        >
          Start Date
        </button>
        <button
          class="ml-2 btn btn-outline-primary"
          type="button"
        >
          End Date
        </button>
        <button
          class="ml-2 btn btn-outline-primary"
          type="button"
        >
          Priority
        </button>
      </div>
    </div>
    <div
      class="fade alert alert-warning show"
      role="alert"
    >
      No task available for selected project, please select another project from the list
    </div>
  </div>
</div>
`;
